### ISSUE TITLE
dynamic image dimensions

### ISSUE THREAD

**Issue by digitalkaoz** [dimsemenov/PhotoSwipe#741]
as you said in the docs its not possible, i even tried and went half way through...just one tiny piece missing. here my attempt:

```js
        var images = [
              { src: "http://foo.com/bar.gif", w:0, h:0}
        ];
        var pswpElement = document.querySelectorAll('.pswp')[0];

        var options = {
            index: 0,
            showHideOpacity: true
        };

        var gallery = new PhotoSwipe(pswpElement, PhotoSwipeUI_Default, images, options);

        gallery.listen('imageLoadComplete', function(index, item) {
            var img = new Image();

            item.w = img.width;
            item.h = img.height;
        });

        gallery.init();
```

sadly it only works good for the lazyloaded images "left" and "right" but not for the current one requested.
in my use case i cant encode the size into the filename as the images are thumbnailed to fit in a box (e.g. image 760+420 should fit in a box of 400x250).
i dont want to make backend requests to get the size, thats not performant. as the example above is half way working i think its possible by PhotoSwipe itself (and it should be a feature anyway ;) )

i think the easiest fix would be to start all displaying stuff after the event is done?

**Comment by dimsemenov** [comment]
In documentation there is an example on how to add or edit slides dynamically, see http://photoswipe.com/documentation/api.html (at the bottom).

You may preload image, retrieve its size, and append (or replace) the dummy image.

**Comment by lucretiu** [comment]
Is there any safe way to do this, so that the user sees no clipping or the image missing? I tried to implement it too, but replacing the dummy image works only when the image was previously loaded and the user hasn't reached that slide.
If the user scrolls fast through the photos, PhotoSwipe will not have time to replace the current slide's width and height, because the preloaded image didn't load yet.
Is there any way to reload the current slide? I just need to change the width and height of the current item.
I have been banging my head against the wall to find a solution for this for hours and hours and hours on end...

**Comment by vralle** [comment]
May be small questions for code example.
What about memory use after nn* click and new images in memory?
Whether to the variable 'img =null;'?

**Comment by Juni4567** [comment]
I solved the issue by adding the following css at the end of my stylesheet

```
.pswp__img {
  height: auto !important;
}
```
and then in the js files where it has the following code I asked it to calculate the width and height of the window and fit our image inside.

```
for(var i = 0; i < numNodes; i++) {

            figureEl = thumbElements[i]; // <figure> element

            // include only element nodes
            if(figureEl.nodeType !== 1) {
            	continue;
            }

            linkEl = figureEl.children[0]; // <a> element

            **//Instead of asking for height and width please calculate it for us We are lazy ;) Juni

            var j_width = $(window).width();
            var j_height = $(window).height();
            size = [j_width, j_height]
**
            // create slide object
            item = {
            	src: linkEl.getAttribute('href'),
            	w: parseInt(size[0], 10),
            	h: parseInt(size[1], 10)
            };



            if(figureEl.children.length > 1) {
                // <figcaption> content
                item.title = figureEl.children[1].innerHTML;
            }

            if(linkEl.children.length > 0) {
                // <img> thumbnail element, retrieving thumbnail url
                item.msrc = linkEl.children[0].getAttribute('src');
            }

            item.el = figureEl; // save link to element for getThumbBoundsFn

            items.push(item);
        }
```

**Comment by denjones** [comment]
I am using PhotoSwipe in my [hexo theme](https://github.com/denjones/hexo-theme-chan) with or without predefined dimension for photos. Here is my approach:

```js
    // Pass data to PhotoSwipe and initialize it
    gallery = new PhotoSwipe(pswpElement, PhotoSwipeUI_Default, items, options);

    gallery.listen('imageLoadComplete', function(index, item) {
      var linkEl = item.el.children[0];
      var img = item.container.children[0];
      if (!linkEl.getAttribute('data-size')) {
        linkEl.setAttribute('data-size', img.naturalWidth + 'x' + img.naturalHeight);
        item.w = img.naturalWidth;
        item.h = img.naturalHeight;
        gallery.invalidateCurrItems();
        gallery.updateSize(true);
      }
    });

    gallery.init();
```

**Comment by redefinered** [comment]
You can easily do this by enclosing the code inside a function and load the dimensions from data attributes when photoswipe is called into view like so:

```
galleryTrigger = $('.pswp-call');

    openPhotoSwipe = function( img, index ) {

        var pswpElement = document.querySelectorAll('.pswp')[0];

        // build items array

        var items = [];

        x = 0;
        galleryTrigger.each(function(){
            items.push({
                src     : $(this).data('url'),
                w       : $(this).data('width'),
                h       : $(this).data('height'),
                title   : $(this).data('caption'),
            });   
            x++;
        })

        // define options if needed
        var options = {
            index               : index,
            showHideOpacity     : true,
            
            // buttons
            closeEl             : true,
            captionEl           : true,
            fullscreenEl        : true,
            zoomEl              : true,
            shareEl             : true,
            counterEl           : true,
            arrowEl             : true,
            preloaderEl         : true,
        }

        var gallery = new PhotoSwipe(pswpElement, PhotoSwipeUI_Default, items, options);

        gallery.init();

    }

    galleryTrigger.click(function(){
        
        var img = $(this).data('url');
        var index = $(this).data('index');

        openPhotoSwipe( img, index );
        
    });
```

**Comment by a-komarev** [comment]
@denjones thanks for the sharing solution. But `naturalWidth` and `naturalHeight` are `undefined` in my case. I've updated your solution a bit:

```js
gallery.listen('imageLoadComplete', function (index, item) {
    var linkEl = item.el.children[0];

    if (!linkEl.getAttribute('data-size') || linkEl.getAttribute('data-size') == 'auto') {
        var img = new Image();
        img.src = linkEl.getAttribute('href');

        linkEl.setAttribute('data-size', img.naturalWidth + 'x' + img.naturalHeight);
        item.w = img.naturalWidth;
        item.h = img.naturalHeight;
        gallery.invalidateCurrItems();
        gallery.updateSize(true);
    }
});
```

**Comment by denjones** [comment]
@a-komarev maybe you should reference an existing img element rather then creating one dynamically.

**Comment by a-komarev** [comment]
@denjones but I have only thumbnails on the page wrapped in `<a>` with href of original images.

**Comment by inigol** [comment]
The photogallery can be opened only one time in Android but works well in iOS, why? everyone can help me?

**Comment by TwentyOneSolutions** [comment]
@denjones Are you sure that this is working for you?  I can't get the `<img>` at that part of the code.  Only after the `imageLoadComplete` event handler exits I see it there.  What does your markup look like for each slide?

I will try @a-komarev 's approach.

**Comment by TwentyOneSolutions** [comment]
Yes, I saw it, but when I tried that technique with many (about 40) images, and with disabling the browser's cache, I was getting errors.

I ended up with the following solution:

first, I added a boolean variable to each slide specifying if `doGetSlideDimensions` is required, so if there is no `data-size` it is set to true.

then, I wrote a function to get the image dimensions for a given slide, like so:

	function getSlideDimensions(slide) {

		if (!slide.doGetSlideDimensions)
			return;

		slide.doGetSlideDimensions = false;

		var img = new Image();
		img.src = slide.src;

		slide.w = img.naturalWidth;
		slide.h = img.naturalHeight;

		photoSwipe.invalidateCurrItems();
		photoSwipe.updateSize(true);
	}

I'm using here a global variable named `photoSwipe` which points to the PhotoSwipe object, but it can be passed as a variable instead, or written as a function inside the initializer where there is a reference to that instance.

The last thing is to call that function from two event handlers, `gettingData` and `imageLoadComplete`: 

	photoSwipe.listen("gettingData", function(index, slide){

		if (slide.doGetSlideDimensions) {

			setTimeout(
				// use setTimeout so that it runs in the event loop
				 function(){ getSlideDimensions(slide); }
				,300
			);
		}
	});

	photoSwipe.listen("imageLoadComplete", function(index, slide){

		if (slide.doGetSlideDimensions)
			getSlideDimensions(slide);
	});

**Comment by elbasankadrija** [comment]
Anyone found a fix for this for us who have thumbnails?

**Comment by TwentyOneSolutions** [comment]
@elbasankadrija my solution works well.  see above.

**Comment by Jogai** [comment]
Can isapi's script behavior made default within photoswipe?

**Comment by karboom** [comment]
@isapi Your solution works for me, but img blinks when it loaded on Android and IOS

**Comment by danbo** [comment]
I missed this thread while trying to figure out the dynamic sizing issue :(

Good news is I may be on to something.

I tested in Firefox on my android and it works quite well. The odd time it gets confused a bit with sizing when I reorient on the fly but I'm willing to live with that for now

```
// note: photoswipe.js: don't forget to move "_shout('imageLoadComplete', index, item);" before "item.img = null;"
                    gallery.listen('imageLoadComplete', 
                        function(index, item) {

                            var actualImageWidth = item.img.naturalWidth;
                            var actualImageHeight = item.img.naturalHeight;
                            
                            // recalculate the fit ratio based on actual image dimensions instead of those which would have been specified in image collection
                            // adapted from photoswipe.js:2743
                            var viewportSize = gallery.viewportSize;
                            var availableX = viewportSize.x;
			                var availableY = viewportSize.y - item.vGap.top - item.vGap.bottom;

                            var hRatio = availableX / actualImageWidth;
				            var vRatio = availableY / actualImageHeight;

				            item.fitRatio = hRatio < vRatio ? hRatio : vRatio;
                            
                            item.w = Math.round(actualImageWidth * item.fitRatio),
                            item.h = Math.round(actualImageHeight * item.fitRatio),
                            
                            gallery.updateSize();
                    });
```

**Comment by DanielHeath** [comment]
I've figured out an approach that should work in all supported browsers. Instead of using an `img` tag to render images, use the following:

          <svg
            class="pswp__img"
            width="${item.w}"
            height="${item.h}"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink"
          >
            <image
              xlink:href="${item.src}"
              x="0"
              y="0"
              height="100"
              width="100"
            />
          </svg>

This lets you specify the width/height you want the photoswipe container to have; if the image has different dimensions, its larger dimension will fill the container and the smaller one is centred.

SVG support is quite good - IE9+, android 3+ etc.

Have I missed something important or does this enable fluid sized images?

 * [ ] Progressive loading - may have to create an `img` tag and not display it to keep the 'is loaded' check
 * [ ] Stretched placeholder - SVG stretches images really nicely
 * [ ] Initial zoom-in transition - SVG has a fixed size and animates nicely
 * [ ] Panning / Zooming - SVG has a fixed size and animates nicely
 * [ ] Caption positioning: I'm not clear from reading the code how this is used, but a fixed size element should keep working anyways.

**Comment by DanielHeath** [comment]
It's not a small job and I certainly wouldn't expect @dimsemenov to do it for us. 

I *would* be keen to hear whether a patch would be welcome or not before attempting it. If not, I might still do it on a fork, but would have to really think about it. @piernik / @isapir hit me up if you're thinking about taking it on yourself, happy to talk about it more.

**Comment by isapir** [comment]
@DanielHeath I have submitted a couple of PRs a while back but it doesn't seem like anyone accepts them:
https://github.com/dimsemenov/PhotoSwipe/pull/1178
https://github.com/dimsemenov/PhotoSwipe/pull/1179

**Comment by DanielHeath** [comment]
It's looking like he's taking a well-deserved holiday (no facebook/twitter/github activity). Hopefully he's having a great time!

**Comment by dimsemenov** [comment]
The feature to dynamically specify dimensions will unlikely be added to core. I specially do not remove  PRs related to it, so people can use them as a start if needed. As [stated in FAQ](http://photoswipe.com/documentation/faq.html#image-size): it's recommended to find a way to retrieve and store dimensions, or use another gallery script that is more suitable for such task.

@isapir, The same goes for "moving" animation when switching slides. There is a performance issue when trying to move large images on a weak PCs, as well as UX issue - a large part of users dislike very large and long moving transition - they just want to view photos.

Though the "fading" transition is currently in development, as it's much more visually elegant and you can pre-render the next view for ~60ms, you can play with a prototype here http://codepen.io/dimsemenov/pen/zNNaNy . 

Along with the feature that allows opening in a zoomed state & panning with mouse movement http://codepen.io/dimsemenov/pen/jrpwEZ

**Comment by piernik** [comment]
@dimsemenov Why not making other version/branch that is more advanced (with calculating dimensions, better animations, even with defined gallery html) so it would be as easy as different lightboxes. If someone is performace geek he would use this library - everyone else new branch/library?

Your library is very good but very complicated to use right away...

**Comment by dimsemenov** [comment]
> Why not making other version/branch that is more advanced (with calculating dimensions, better animations, even with defined gallery html)

I don't have resources to maintain two separate versions of the script.

> Your library is very good but very complicated to use right away...

I agree with that, that's why there will be a module (or even a separate file) that would allow easier initialization directly from list of thumbnails, take a look at what markup is required for this example - http://codepen.io/dimsemenov/pen/zNNaNy Though it's still in development.

**Comment by piernik** [comment]
Oh thats great. Will the new module has also embed html of gallery? Whe do You plan to release? Should I wait or search for other lightbox?

And why not adding official module for dynamic diamentions?

**Comment by isapir** [comment]
That's fine by me.  I just don't want to create a PR that will be forever lost in _GithubSpace_.

Also, I've concluded that for my projects I will use @dimsemenov 's Royal Slider (http://dimsemenov.com/plugins/royal-slider/).  Seems to be well worth its affordable price.

**Comment by DanielHeath** [comment]
I'd love to see a feature comparison between the two; $WORK are generally OK to pay for stuff but I'd have to be able to write up why we should use the paid one over the free one.

**Comment by piernik** [comment]
@dimsemenov When You plan to release new version?

**Comment by piernik** [comment]
@dimsemenov Can You give me some timeline? I want to use PhotoSwipe in my project, but if You won't give it upgrade until march I have to look for other solution.

**Comment by DanielHeath** [comment]
The entitlement on display here is staggering.

This is free stuff given to you by volunteers. Pay your share or quit whinging.

**Comment by piernik** [comment]
@DanielHeath I have no problem with paying for good solutions. Even now I have paid lightbox. But I want to switch and I liked PhotoSwipe. If @dimsemenov, as he mentioned,  has in plans to release new version I want to know timeline (not exact).

**Comment by windowsair** [comment]
@denjones Thank you very much. I tried your method, but found that after closing the panel and re-opened it, there will not be able to resolve the picture size. Does my code seem to have problems?
```javascript
var initPhotoSwipeFromDOM = function(gallerySelector) {

    // 解析来自DOM元素幻灯片数据（URL，标题，大小...）
    // (children of gallerySelector)
    var parseThumbnailElements = function(el) {
        var thumbElements = el.childNodes,
            numNodes = thumbElements.length,
            items = [],
            figureEl,
            linkEl,
            size,
            item,
            divEl;

        for(var i = 0; i < numNodes; i++) {

            figureEl = thumbElements[i]; // <figure> element

            // 仅包括元素节点
            if(figureEl.nodeType !== 1) {
                continue;
            }
            divEl = figureEl.children[0];
            linkEl = divEl.children[0]; // <a> element
            
            size = linkEl.getAttribute('data-size').split('x');

            // 创建幻灯片对象
            item = {
                        src: linkEl.getAttribute('href'),
                        w: parseInt(size[0], 10),
                        h: parseInt(size[1], 10),
                        author: linkEl.getAttribute('data-author')
                    };



            if(figureEl.children.length > 1) {
                // <figcaption> content
                item.title = figureEl.children[1].innerHTML; 
            }

            if(linkEl.children.length > 0) {
                // <img> 缩略图节点, 检索缩略图网址
                item.msrc = linkEl.children[0].getAttribute('src');
            } 

            item.el = figureEl; // 保存链接元素 for getThumbBoundsFn
            items.push(item);
        }

        return items;
    };

    // 查找最近的父节点
    var closest = function closest(el, fn) {
        return el && ( fn(el) ? el : closest(el.parentNode, fn) );
    };

    // 当用户点击缩略图触发
    var onThumbnailsClick = function(e) {
        e = e || window.event;
        e.preventDefault ? e.preventDefault() : e.returnValue = false;

        var eTarget = e.target || e.srcElement;

        // find root element of slide
        var clickedListItem = closest(eTarget, function(el) {
            return (el.tagName && el.tagName.toUpperCase() === 'FIGURE');
        });

        if(!clickedListItem) {
            return;
        }

        // find index of clicked item by looping through all child nodes
        // alternatively, you may define index via data- attribute
        var clickedGallery = clickedListItem.parentNode,
            childNodes = clickedListItem.parentNode.childNodes,
            numChildNodes = childNodes.length,
            nodeIndex = 0,
            index;

        for (var i = 0; i < numChildNodes; i++) {
            if(childNodes[i].nodeType !== 1) { 
                continue; 
            }

            if(childNodes[i] === clickedListItem) {
                index = nodeIndex;
                break;
            }
            nodeIndex++;
        }



        if(index >= 0) {
            // open PhotoSwipe if valid index found
            openPhotoSwipe( index, clickedGallery );
        }
        return false;
    };

    // parse picture index and gallery index from URL (#&pid=1&gid=2)
    var photoswipeParseHash = function() {
        var hash = window.location.hash.substring(1),
        params = {};

        if(hash.length < 5) {
            return params;
        }

        var vars = hash.split('&');
        for (var i = 0; i < vars.length; i++) {
            if(!vars[i]) {
                continue;
            }
            var pair = vars[i].split('=');  
            if(pair.length < 2) {
                continue;
            }           
            params[pair[0]] = pair[1];
        }

        if(params.gid) {
            params.gid = parseInt(params.gid, 10);
        }

        return params;
    };

    var openPhotoSwipe = function(index, galleryElement, disableAnimation, fromURL) {
        var pswpElement = document.querySelectorAll('.pswp')[0],
            gallery,
            options,
            items;

        items = parseThumbnailElements(galleryElement);

        // 这里可以定义参数
        options = {

                    openZoomed: true,
  history: false,
  mouseMovePan: true,


                    galleryUID: galleryElement.getAttribute('data-pswp-uid'),

                    getThumbBoundsFn: function(index) {
                        // See Options->getThumbBoundsFn section of docs for more info
                        var thumbnail = items[index].el.children[0],
                            pageYScroll = window.pageYOffset || document.documentElement.scrollTop,
                            rect = thumbnail.getBoundingClientRect(); 

                        return {x:rect.left, y:rect.top + pageYScroll, w:rect.width};
                    },

                    addCaptionHTMLFn: function(item, captionEl, isFake) {
                        if(!item.title) {
                            captionEl.children[0].innerText = '';
                            return false;
                        }
                        captionEl.children[0].innerHTML = item.title +  '<br/><small>Photo: ' + item.author + '</small>';
                        return true;
                    },
                    
                };


        // PhotoSwipe opened from URL
        if(fromURL) {
            if(options.galleryPIDs) {
                // parse real index when custom PIDs are used 
                for(var j = 0; j < items.length; j++) {
                    if(items[j].pid == index) {
                        options.index = j;
                        break;
                    }
                }
            } else {
                // in URL indexes start from 1
                options.index = parseInt(index, 10) - 1;
            }
        } else {
            options.index = parseInt(index, 10);
        }

        // exit if index not found
        if( isNaN(options.index) ) {
            return;
        }

        if(disableAnimation) {
            options.showAnimationDuration = 0;
        }

        // Pass data to PhotoSwipe and initialize it
        gallery = new PhotoSwipe( pswpElement, PhotoSwipeUI_Default, items, options);
         gallery.listen('imageLoadComplete', function(index, item) {
      var linkEl = item.el.children[0];
      var img = item.container.children[0];
      if (!linkEl.getAttribute('data-size')) {
        linkEl.setAttribute('data-size', img.naturalWidth + 'x' + img.naturalHeight);
        item.w = img.naturalWidth;
        item.h = img.naturalHeight;
        gallery.invalidateCurrItems();
        gallery.updateSize(true);
      }
    });
        gallery.init();
    };

    // loop through all gallery elements and bind events
    var galleryElements = document.querySelectorAll( gallerySelector );

    for(var i = 0, l = galleryElements.length; i < l; i++) {
        galleryElements[i].setAttribute('data-pswp-uid', i+1);
        galleryElements[i].onclick = onThumbnailsClick;
    }

    // Parse URL and open gallery if it contains #&pid=3&gid=1
    var hashData = photoswipeParseHash();
    if(hashData.pid && hashData.gid) {
        openPhotoSwipe( hashData.pid ,  galleryElements[ hashData.gid - 1 ], true, true );
    }
    };

    // execute above function
    initPhotoSwipeFromDOM('.my-gallery');

    
    $(".my-gallery>figure>div").each(function(){
        $(this).height($(this).width());
    });
```

**Comment by tscislo-lingaro** [comment]
@denjones I like your solution, but when we put in onLoad handler those two lines: 
                this.photoswipe.invalidateCurrItems();
                this.photoswipe.updateSize(true)

there is a nasty side effect in that current photo that we zoom in gets reset when the NEXT photo loads. This is really annoying. Any ideas on how to solve this?

**Comment by andrerfneves** [comment]
Just passing by to say @isapir's solution worked for me.

**Comment by tscislo** [comment]
@isapir @andrerfneves 
It works fine, but there is blink effect of currently visible photo already mentioned by @karboom It happens when we trigger:
this.photoswipe.invalidateCurrItems();
this.photoswipe.updateSize(true);
Any ideas on how to prevent this blink, which affects currently visible photo?

**Comment by stucki** [comment]
@tscislo Did you already try the alternative solution from #796?

**Comment by tscislo-lingaro** [comment]
@stucki 
Thanks! Unfortunately even if I removed invalidateCurrItems() I still get flickering.

**Comment by JefferyHus** [comment]
presented solutions in this issue are not working, in my case as I only wanted the plugin to zoom one and only image after click:

```xml
<img src="..." onclick="zoomMe(this)" />
```

```javascript
// get demensions
function getImageDemensions(imageSrc)
{
    let image = new Image()
    return new Promise(function(resolve, reject) {
        image.src = imageSrc
        image.addEventListener('load', function () {
            resolve(image)
        }, false)
    })
}
....
getImageDemensions(imageOriginal)
.then(function (result) {
    // add a new item
   window.pswpItems = [{
   src: imageOriginal,
   w: result.naturalWidth,
   h: result.naturalHeight}]
   // View one image
   window.pswp = new 
   PhotoSwipe(window.pswpElements,PhotoSwipeUI_Default,window.pswpItems,window.pswpOptions)
       // start
       window.pswp.init()
   })
...
```

This solution works without an issue.

P.S: you can add a reject if you want to the promise, as example check if the src is empty, image resolution is not correct...etc

Goodluck

**Comment by iamgreyy** [comment]
I was fixed it by normalizing my image array as

```
  prepareImagesToPreload = images => {
    return images.map(({ w, h, ...each }) => ({
      w: 0,
      h: 0,
      ...each
    }))
  }
```

Then i add "beforeChange" listener and do this 
```
beforeChange={(instance) => {
            const {
              currItem: { src }
            } = instance

            const image = new Image()
            image.src = src

            image.onload = () => {
              instance.currItem.w = image.width
              instance.currItem.h = image.height

              instance.updateSize()
            }
          }}
```

Next i set loader to photoswipe wrapper as background-image property:
```
.pswp__item {
  background: transparent url('./loader.svg') center no-repeat;
}

```

It works fine for me!

**Comment by amk9889** [comment]
try this
`
var items = [
		    {
		        src: image.src,
		        w: image.width,
		        h: image.height
		    }
		];
		var options = {
		    index: 0 
		};
		
		var gallery = new PhotoSwipe( pswpElement, PhotoSwipeUI_Default, items, options);


`

**Comment by borconi** [comment]
Thanks to @antonkomarev  for the excellent idea. However I discovered that in some cases the image element is null, or the whole element is null, so I made a slight tweak which seems to work a treat.


 ```
var imagelist=[
                        {src:'www.example.com/example1.jpg',w:null,h:null},
                        {src:'www.example.com/example2.jpg',w:null,h:null},
                        {src:'www.example.com/example3.jpg',w:null,h:null},
                        {src:'www.example.com/example4.jpg',w:null,h:null}
                       ]
var pswpElement = document.querySelectorAll('.pswp')[0];
 var gallery = new PhotoSwipe( pswpElement, PhotoSwipeUI_Default, imagelist);
 gallery.listen('imageLoadComplete', function(index, item) { 
      if (item.w==null) {
	var img = new Image();
	img.src=item.src;
        item.w = img.naturalWidth;
        item.h = img.naturalHeight;
        gallery.invalidateCurrItems();
        gallery.updateSize(true);
      }
    });
gallery.init();
```

**Comment by buff2017** [comment]
i tried like this , its working for me
```javascript
    // .......
        let gallery = new PhotoSwipe(this.$pswpElement, PhotoSwipeUI_Default, items, options)
        gallery.listen('imageLoadComplete', function (index, item) {
          if (item.h < 1 || item.w < 1) {
            let img = new Image()
            img.onload = () => {
              item.w = img.width
              item.h = img.height
              gallery.invalidateCurrItems()
              gallery.updateSize(true)
            }
            img.src = item.src
          }
        })
        gallery.init()
```

**Comment by djekanovic** [comment]
Here is a solution (still not perfect but it works for me) that i created using only css

.pswp img {
    max-height: 100vh !important;
    width: auto !important;
position: static !important;
}

.pswp__zoom-wrap {
    display: flex !important;
    justify-content: center;
    align-items: center;
    transform: none !important;
}

**Comment by dikshantpatodia** [comment]
This solves the issue, retaining the transition and transformation.

```
.pswp img {
  max-width: none;
  object-fit: contain;
}
```

**Comment by hykilpikonna** [comment]
## Perfect solution so far:

**Thanks to @dikshantpatodia, this solution does not require JS nor knowing the image size beforehand.**

> This solves the issue, retaining the transition and transformation.
> 
> ```
> .pswp img {
>   max-width: none;
>   object-fit: contain;
> }
> ```

However, to achieve the full-screen effect, `w` and `h` in the slides must be set to `windpw.innerWidth` and `window.innerHeight`. This way you can keep both the proportions and display the image in full screen without knowing the ratio of the image.

Eg.

```js
var slides = [
    {
        src: 'path/to/image1.jpg', // path to image
        title: 'Image Caption'
        w: window.innerWidth, // image width
        h: window.innerHeight, // image height
    }
];
```

### Screenshots:

When they are set to `w: 0, h: 0`

![image](https://user-images.githubusercontent.com/22280294/71448544-32e1be00-270b-11ea-987c-326769189dbb.png)

When they are set to values with defined ratio (eg. `w: 1000, h: 1000`)

![image](https://user-images.githubusercontent.com/22280294/71448584-de8b0e00-270b-11ea-8e9d-5e6d663ff37a.png)

When they are set to `w: window.innerWidth, h: window.innerHeight`

![image](https://user-images.githubusercontent.com/22280294/71448604-16925100-270c-11ea-9d67-16b5f1b56641.png)

Also works responsively:

![image](https://user-images.githubusercontent.com/22280294/71448620-4b060d00-270c-11ea-9fe8-cb6b6fbc4863.png)

**Comment by dikshantpatodia** [comment]
Thanks @hykilpikonna :)

**Comment by darko-r** [comment]
The way I did it was by calculating all gallery images sizes and then write that info in attribute. It works perfectly for me, for six big images (bigger than 2000px wide) it takes ~30ms. Execute the code after DOMContentLoaded event.

`setTimeout(function() {    // Wait for empty stack
    const galleryLinks = document.querySelectorAll('.image-link-class');    // Get all links
    galleryLinks.forEach((element) => {
        let image = new Image();
        image.src = element.getAttribute('href');
        image.onload = () => {    // Wait for image to load
            element.setAttribute('data-size', `${image.naturalWidth}x${image.naturalHeight}`);
            image = null;    // Free resource so garbage collector can clean it
        }
    });
}, 1);`

**Comment by phegman** [comment]
https://github.com/dimsemenov/PhotoSwipe/issues/741#issuecomment-568915867 works well for me but I think you would want to be carful about cleaning up the `resize` event listener. I also used lodash `throttle` so the `resize` event listener isn't fired too many times.

```js
const gallery = new PhotoswipeJS(
  pswpElement,
  PhotoSwipeUI_Default,
  items.map(({ src, thumbnailSrc }) => ({
    src,
    msrc: thumbnailSrc,
    w: window.innerWidth,
    h: window.innerHeight,
  })),
  options
);

gallery.init();

const updateItemsDimensions = (items: PhotoswipeItem[]) => {
  items.forEach((item) => {
    item.w = window.innerWidth;
    item.h = window.innerHeight;
  });
};

const updateItemsDimensionsWithBinding = throttle(
  updateItemsDimensions.bind(this, gallery.items),
  200
);

window.addEventListener("resize", updateItemsDimensionsWithBinding);

gallery.listen("destroy", () => {
  window.removeEventListener("resize", updateItemsDimensionsWithBinding);
});
```

**Comment by dotterian** [comment]
Actually baffled to see this issue.

It took me one day of work and a handful of asyncs, awaits and Promises to add image preloading and loading indicator on thumbnail and add TypeScript types to your code. Still didn't implement loading indicator on slide change but it won't be that hard, when I'll have some time and get to this issue.

Sadly, I can't share my code (at least for now) for several reasons:
1. It's a fork, not a "filter" or eventListener that you can add right away, so it should be either merged into this repo, published on npm as separate package or user should move whole `src` directory to their code manually and bundle it with their bundler (it's an option that I use right now since it's just simpler for me to do things this way).
2. It's not "open-source ready" and is just a hacky patch without any customization and styling options or foolproofing and may cause problems with some plugins.
3. It doesn't look good at this point and needs at least another day/two of work which I, for now, don't have and don't know when I'll be able to finish it

Probably I'll publish it someday, but for now I'll just ask this one question to maintainer:
Why on earth is this still an issue after 7 years of development?

It's not so hard to implement. Your code architecture allows this to be implemented fairly easy if you'd ever took your time to do this. And I don't see any reason why you shouldn't implement this. At this point I'm frustrated that I had to fork you library to add this feature. Imho, implementing drag'n'drop pan and multitouch zoom with spring animations is much-much harder than this "conundrum", that, somehow, isn't solved to this day.

And in the end of this "angry" post I'll add this to sweeten the pill:
Actually your library is great and I pretty much like it *(and I actually did read all of the code in `src` folder since I had to add proper types to it)*. Thanks for you work.

**Comment by Jogai** [comment]
@dotterian at least open a PR to add the ts types. That way at least part of your work wasn't for nothing.

**Comment by markfinst** [comment]
@dotterian It's a collaborative open source project. Please file a pull request if it's simple to do

**Comment by martin-juul** [comment]
This doesn't seem possible with the new photoswipe 5 library. The first few slides are placed improperly on the screen (all the way to the left), but the next slides are loaded correctly. What gives?

```typescript
    this.lightbox = new PhotoSwipe({
      appendToEl: container,
      dataSource: [{src: 'https://path/to/pic'}],
      index: 0,
    });

    this.lightbox.on('contentLoad', (e) => {
      const { content } = e;
      e.preventDefault();

      content.state = 'loading';

      const img = new Image();
      img.addEventListener('load', () => {
        const { width, height } = this.setSizes(img);

        content.height = height;
        content.width = width;

        content.setDisplayedSize(width, height);

        content.onLoaded();
      });
      img.addEventListener('error', () => {
        content.onError();
      });

      img.className = 'pswp__img';
      img.src = content.data.src;
      img.setAttribute('alt', '');
      content.element = img;
    });

  private setSizes(img: HTMLImageElement): { width: number; height: number } {
    // ...
  }
```

**Comment by kfeng0806** [comment]
```javascript
const imageList = [
    {
        src: ...,
        w: 0,
        h: 0,
    },
    { ... }
];
const pswp = new PhotoSwipe(ELEMENT, UI, imageList, OPTIONS);

pswp.listen('gettingData', (index, item) => updateImageDimention(pswp, index, item));
pswp.listen('imageLoadComplete', (index, item) => updateImageDimention(pswp, index, item));

updateImageDimention(pswpInstance, index, item) {
    if (item.h < 1 || item.w < 1) {
        const image = new Image();

        image.onload = () => {
            item.w = mage.width;
            item.h = iimage.height;
            pswpInstance.updateSize(true);

            imageList[index] = {
                ...imageList[index],
                w: width,
                h: height,
            }
        }

        image.src = item.src;
    }
},
```

Works perfectly for me.

**Comment by opctim** [comment]
This worked for me:

`
.pswp__img {
    object-fit: contain;
    max-height: 100vh !important;
    max-width: 100vw !important;
}
`

**Comment by jerrywu001** [comment]
**v5.x usage**

```tsx
import { useLayoutEffect } from "react";
import PhotoSwipeLightbox from "photoswipe/lightbox";
import 'photoswipe/style.css';

const dataSource = [
  'https://cdn.photoswipe.com/photoswipe-demo-images/photos/1/img-2500.jpg',
  'https://source.unsplash.com/RJzHlbKf6eY/1950x1300',
  'https://cdn.photoswipe.com/photoswipe-demo-images/photos/srcset-test/4-1000x1500.png',
  'https://dummyimage.com/1500x1000/555/fff/?text=1500x1000',
  'https://cdn.photoswipe.com/photoswipe-demo-images/photos/3/img-2500.jpg',
  'https://cdn.photoswipe.com/photoswipe-demo-images/photos/2/img-2500.jpg'
].map(v => ({ src: v }));

export default function App() {
  useLayoutEffect(() => {
    const lightbox = new PhotoSwipeLightbox({
      dataSource,
      pswpModule: () => import('photoswipe'),
    });

    lightbox.init();
    lightbox?.loadAndOpen(0, dataSource);
    
    lightbox.on('beforeOpen', () => {
      const pswp = lightbox.pswp;
      const ds = pswp?.options?.dataSource;
      if (Array.isArray(ds)) {
        for (let idx = 0, len = ds.length; idx < len; idx++) {
          const item = ds[idx];
          const img = new Image();
          img.onload = () => {
            item.width = img.naturalWidth
            item.height = img.naturalHeight
            pswp?.refreshSlideContent(idx)
          };
          img.src = item.src as string;
        }
      }
    });
  }, [])

  return null;
}
```

![image](https://github.com/dimsemenov/PhotoSwipe/assets/12525415/ba6a2af0-420f-4039-951d-5e6a5afc33b3)

**Comment by mrkpatchaa** [comment]
@jerrywu001 I adapted your code a little bit for usage with react-photoswipe-gallery
```
  const onOpen = (pswpInstance) => {
    pswpInstance.on('beforeOpen', () => {
      const ds = pswpInstance?.options?.dataSource
      if (Array.isArray(ds)) {
        for (let idx = 0, len = ds.length; idx < len; idx++) {
          const item = ds[idx]
          const img = new Image()
          img.onload = () => {
            item.width = img.naturalWidth
            item.height = img.naturalHeight
            pswpInstance?.refreshSlideContent(idx)
          }
          img.src = item.src as string
        }
      }
    })
  }

...
<Gallery uiElements={uiElements} onOpen={onOpen}>
...
```

**Comment by bsor-dev** [comment]
> @jerrywu001 I adapted your code a little bit for usage with react-photoswipe-gallery
> 
> ```
>   const onBeforeOpen = (pswpInstance) => {
>     pswpInstance.on('beforeOpen', () => {
>       const ds = pswpInstance?.options?.dataSource
>       if (Array.isArray(ds)) {
>         for (let idx = 0, len = ds.length; idx < len; idx++) {
>           const item = ds[idx]
>           const img = new Image()
>           img.onload = () => {
>             item.width = img.naturalWidth
>             item.height = img.naturalHeight
>             pswpInstance?.refreshSlideContent(idx)
>           }
>           img.src = item.src as string
>         }
>       }
>     })
>   }
> 
> ...
> <Gallery uiElements={uiElements} onBeforeOpen={onBeforeOpen}>
> ...
> ```

Cool

**Comment by fcorneli** [comment]
Next also works:
```javascript
let $this = this;
            this.lightbox = new PhotoSwipeLightbox({
                pswpModule: PhotoSwipe
            });
            this.lightbox.addFilter("numItems", (numItems) => {
                return $this.imageData.length;
            });
            this.lightbox.addFilter("itemData", (itemData, index) => {
                let zoomImage = $this.imageData[index].zoomImage;
                if ($this.imageData[index].width) {
                    return {
                        src: zoomImage,
                        width: $this.imageData[index].width,
                        height: $this.imageData[index].height
                    };
                }
                let image = new Image();
                image.onload = () => {
                    $this.imageData[index].width = image.naturalWidth;
                    $this.imageData[index].height = image.naturalHeight;
                    $this.lightbox.pswp.refreshSlideContent(index);
                };
                image.src = zoomImage;
                return {};
            });
            this.lightbox.init();
```
So if you don't have `width` and `height` you give back `{}` and load the image, after which you can set `width` and `height` and then your simply force a `refreshSlideContent`.

**Comment by pjotrsavitski** [comment]
One more possibility would be to add an event handler to image load event and set width and height as data attributes using the callback. Current case has images of just one size and data attributes would be set to each image once that is loaded. Not sure if it has any unexpected side effects, but the updated data seems to be picked up just fine once slide is being opened. That will not change already opened slides, but that is tolerable in current case.

```html
<a :href="image.url" target="_blank">
    <img :src="image.url" alt="Image" @load="galleryImageLoaded"/>
</a>
```

```js
const galleryImageLoaded = (event) => {
    event.target.parentNode.setAttribute("data-pswp-width", event.target.naturalWidth);
    event.target.parentNode.setAttribute("data-pswp-height", event.target.naturalHeight);
};
```

**Comment by dimsemenov** [comment]
Images with no predefined dimensions will be a supported in v6, more info here https://github.com/dimsemenov/PhotoSwipe/discussions/2170
